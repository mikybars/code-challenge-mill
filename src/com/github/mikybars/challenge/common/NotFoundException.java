package com.github.mikybars.challenge.common;

public class NotFoundException extends RuntimeException {

  public NotFoundException(String message) {
    super(message);
  }
}
