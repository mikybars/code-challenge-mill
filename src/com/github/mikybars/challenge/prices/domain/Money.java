package com.github.mikybars.challenge.prices.domain;

import java.math.BigDecimal;
import java.util.Currency;

public record Money(BigDecimal amount, Currency currency) {

}
