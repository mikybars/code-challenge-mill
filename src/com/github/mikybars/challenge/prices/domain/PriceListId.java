package com.github.mikybars.challenge.prices.domain;

public record PriceListId(String id) {

}
