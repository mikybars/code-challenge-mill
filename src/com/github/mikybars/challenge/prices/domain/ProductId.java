package com.github.mikybars.challenge.prices.domain;

public record ProductId(String id) {

}
