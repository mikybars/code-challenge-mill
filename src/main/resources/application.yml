spring:
  application:
    name: code-challenge
  config:
    import:
      - error-handling.yml