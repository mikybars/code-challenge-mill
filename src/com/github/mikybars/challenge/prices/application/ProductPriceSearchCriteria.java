package com.github.mikybars.challenge.prices.application;

import com.github.mikybars.challenge.prices.domain.ProductId;
import java.time.LocalDateTime;

public record ProductPriceSearchCriteria(LocalDateTime applicationDate, ProductId productId) {

}