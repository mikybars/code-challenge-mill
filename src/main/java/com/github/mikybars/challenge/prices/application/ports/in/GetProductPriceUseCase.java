package com.github.mikybars.challenge.prices.application.ports.in;

import com.github.mikybars.challenge.prices.application.ProductPriceSearchCriteria;
import com.github.mikybars.challenge.prices.domain.ProductPrice;

public interface GetProductPriceUseCase {

  ProductPrice execute(ProductPriceSearchCriteria searchCriteria);
}
