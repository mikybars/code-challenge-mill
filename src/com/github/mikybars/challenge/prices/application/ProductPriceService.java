package com.github.mikybars.challenge.prices.application;

import com.github.mikybars.challenge.common.NotFoundException;
import com.github.mikybars.challenge.prices.application.ports.in.GetProductPriceUseCase;
import com.github.mikybars.challenge.prices.application.ports.out.ProductPriceRepository;
import com.github.mikybars.challenge.prices.domain.ProductPrice;
import java.util.function.Supplier;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
class ProductPriceService implements GetProductPriceUseCase {

  private final ProductPriceRepository productPriceRepository;

  @Override
  public ProductPrice execute(ProductPriceSearchCriteria searchCriteria) {
    return productPriceRepository
        .findTheHighestRankedBy(searchCriteria)
        .orElseThrow(notFoundExceptionTriggeredBy(searchCriteria));
  }

  private static Supplier<NotFoundException> notFoundExceptionTriggeredBy(
      ProductPriceSearchCriteria criteria) {
    return () -> new NotFoundException("no product price found for %s".formatted(criteria));
  }
}
