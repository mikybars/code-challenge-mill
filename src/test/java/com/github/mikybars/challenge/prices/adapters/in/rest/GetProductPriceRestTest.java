package com.github.mikybars.challenge.prices.adapters.in.rest;

import static org.mockito.Mockito.when;

import com.github.mikybars.challenge.common.NotFoundException;
import com.github.mikybars.challenge.prices.ProductPrices;
import com.github.mikybars.challenge.prices.application.ProductPriceSearchCriteria;
import com.github.mikybars.challenge.prices.application.ports.in.GetProductPriceUseCase;
import com.github.mikybars.challenge.prices.domain.ProductId;
import java.time.LocalDateTime;
import org.approvaltests.JsonApprovals;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest;
import org.springframework.context.annotation.Import;
import org.springframework.http.MediaType;
import org.springframework.test.context.bean.override.mockito.MockitoBean;
import org.springframework.test.web.reactive.server.WebTestClient;
import org.springframework.test.web.reactive.server.WebTestClient.ResponseSpec;

@WebMvcTest
@Import(ProductPriceRestMapperImpl.class)
class GetProductPriceRestTest {

  @Autowired
  WebTestClient webClient;

  @MockitoBean
  GetProductPriceUseCase getProductPriceUseCase;

  @Test
  void getProductPrice() {
    when(getProductPriceUseCase.execute(
        new ProductPriceSearchCriteria(sometime(), ProductPrices.any().productId()))
    ).thenReturn(ProductPrices.any());

    ResponseSpec response = webClient.get()
        .uri(uriBuilder -> uriBuilder
            .path("/prices")
            .queryParam("applicationDate", sometime().toString())
            .queryParam("productId", ProductPrices.any().productId().id())
            .build())
        .accept(MediaType.APPLICATION_JSON)
        .exchange();

    response
        .expectStatus().isOk()
        .expectBody(String.class).consumeWith(result ->
            JsonApprovals.verifyJson(result.getResponseBody()));
  }

  @Test
  void doesNotGetProductPrice() {
    when(getProductPriceUseCase.execute(
        new ProductPriceSearchCriteria(sometime(), productNotFoundId()))
    ).thenThrow(new NotFoundException("no price found for input params"));

    ResponseSpec response = webClient.get()
        .uri(uriBuilder -> uriBuilder
            .path("/prices")
            .queryParam("applicationDate", sometime().toString())
            .queryParam("productId", productNotFoundId().id())
            .build())
        .accept(MediaType.APPLICATION_JSON)
        .exchange();

    response
        .expectStatus().isNotFound()
        .expectBody(String.class).consumeWith(result ->
            JsonApprovals.verifyJson(result.getResponseBody()));
  }

  @Test
  void validationErrorsMeetStandards() {
    ResponseSpec response = webClient.get()
        .uri(uriBuilder -> uriBuilder
            .path("/prices")
            // .queryParam("applicationDate", "is missing")
            .build())
        .accept(MediaType.APPLICATION_JSON)
        .exchange();

    response
        .expectStatus().isBadRequest()
        .expectBody()
        .jsonPath("code").exists()
        .jsonPath("message").exists();
  }

  static LocalDateTime sometime() {
    return LocalDateTime.parse("2020-06-14T00:00:00");
  }

  static ProductId productNotFoundId() {
    return new ProductId("35456");
  }
}