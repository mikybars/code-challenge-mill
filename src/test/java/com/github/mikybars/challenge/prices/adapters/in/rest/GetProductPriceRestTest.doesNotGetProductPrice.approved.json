{
  "code": "NOT_FOUND",
  "message": "no price found for input params"
}